### PR TITLE
Add the ability to specify resource estimates for queries

### PR DESCRIPTION
**Pull request by raghavsethi**
cc @highker

### CODE REVIEW

**Comment by findepi** [comment]
What's the intended usage of resource estimates? Is it like _hints_?

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java [hunk 1]
```diff
@@ -260,6 +271,39 @@ public boolean supportClientTransaction()
...
+            long value;
+            try {
+                value = Long.parseLong(nameValue.get(1));
+            }
+            catch (NumberFormatException ignored) {
+                throw badRequest(format("Expected integral value for resource estimate %s", name));
+            }
+
+            // @reviewer: Should I do this or do switch (foo.toUpperCase()) and also uppercase all the static finals
+            if (name.equalsIgnoreCase(ResourceEstimates.EXECUTION_TIME_MILLIS)) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java]: Let's use underscore so it's consistent with session properties

File: presto-spi/src/main/java/com/facebook/presto/spi/session/ResourceEstimates.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Estimated resource usage for a query.
+ *
+ * This class is under active development and should be considered beta.
+ */
+public final class ResourceEstimates
+{
+    public static final String EXECUTION_TIME_MILLIS = "executionTimeMillis";
```
Comment (electrum) [inline presto-spi/src/main/java/com/facebook/presto/spi/session/ResourceEstimates.java]: Inline these

**Comment by electrum** [comment]
One thing this doesn't allow is execution of a SQL script like `presto -f foo.sql` where each query has different estimates:
```
begin;

create table ...;

set resource estimate peak_memory = '1TB';
insert into ...;

set resource estimate peak_memory = '3TB';
insert into ...;

commit;
```

**Comment by raghavsethi** [comment]
@findepi Yes these are like hints that the resource group manager might take into account. We're not completely certain how this will be used but wanted to get some early feedback on getting a client-provided estimate.

File: presto-cli/src/main/java/com/facebook/presto/cli/ClientOptions.java [hunk 3]
```diff
@@ -192,9 +200,77 @@ public static URI parseServer(String server)
...
+
+        public ClientResourceEstimate(String resourceEstimate)
+        {
+            List<String> nameValue = NAME_VALUE_SPLITTER.splitToList(resourceEstimate);
+            checkArgument(nameValue.size() == 2, "Resource estimate: %s", resourceEstimate);
+
+            this.resource = nameValue.get(0);
+            this.estimate = nameValue.get(1);
+            checkArgument(ASCII_ENCODER.canEncode(resource), "Resource is not US_ASCII: %s", resource);
+            checkArgument(resource.indexOf('=') < 0, "Resource must not contain '=': %s", resource);
```
Comment (dain) [inline presto-cli/src/main/java/com/facebook/presto/cli/ClientOptions.java]: Do you want to ensure that the key and value are not empty strings?

File: presto-cli/src/main/java/com/facebook/presto/cli/ClientOptions.java [hunk 4]
```diff
@@ -192,9 +200,77 @@ public static URI parseServer(String server)
...
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            ClientResourceEstimate estimate1 = (ClientResourceEstimate) o;
```
Comment (dain) [inline presto-cli/src/main/java/com/facebook/presto/cli/ClientOptions.java]: `estimate1`?

File: presto-client/src/main/java/com/facebook/presto/client/ClientSession.java [hunk 5]
```diff
@@ -91,8 +94,15 @@ public ClientSession(
             checkArgument(!clientTag.contains(","), "client tag cannot contain ','");
         }
 
-        // verify the properties are valid
+        // verify that resource estimates are valid
         CharsetEncoder charsetEncoder = US_ASCII.newEncoder();
+        for (Entry<String, String> entry : resourceEstimates.entrySet()) {
+            checkArgument(!entry.getKey().isEmpty(), "Resource name is empty");
+            checkArgument(entry.getKey().indexOf('=') < 0, "Resource name must not contain '=': %s", entry.getKey());
+            checkArgument(charsetEncoder.canEncode(entry.getKey()), "Resource name is not US_ASCII: %s", entry.getKey());
```
Comment (dain) [inline presto-client/src/main/java/com/facebook/presto/client/ClientSession.java]: is empty value allowed?  Do you want to check the encoding of the value?

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java [hunk 6]
```diff
@@ -260,6 +273,33 @@ public boolean supportClientTransaction()
...
+        ResourceEstimateBuilder builder = new ResourceEstimateBuilder();
+        for (String header : splitSessionHeader(servletRequest.getHeaders(PRESTO_RESOURCE_ESTIMATE))) {
+            List<String> nameValue = Splitter.on('=').limit(2).trimResults().splitToList(header);
+            assertRequest(nameValue.size() == 2, "Invalid %s header", PRESTO_RESOURCE_ESTIMATE);
+            String name = nameValue.get(0);
+            String value = nameValue.get(1);
+
+            switch(name.toUpperCase()) {
+                case ResourceEstimates.EXECUTION_TIME:
+                    builder.setExecutionTime(Duration.valueOf(value));
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java]: Catch the `IllegalArguentException` from these and throw a badRequest?

File: presto-main/src/main/java/com/facebook/presto/server/SessionContext.java [hunk 7]
```diff
@@ -45,6 +46,9 @@
 
     Set<String> getClientTags();
 
+    @Nullable
+    ResourceEstimates getResourceEstimates();
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/server/SessionContext.java]: I don't see the usage, but please verify this is actually nullable.

File: presto-spi/src/main/java/com/facebook/presto/spi/session/ResourceEstimates.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        return cpuTime;
+    }
+
+    @JsonProperty
+    public Optional<DataSize> getPeakMemory()
+    {
+        return peakMemory;
+    }
+}
```
Comment (dain) [inline presto-spi/src/main/java/com/facebook/presto/spi/session/ResourceEstimates.java]: Maybe add a `toString`

File: presto-main/src/main/java/com/facebook/presto/SessionRepresentation.java [hunk 9]
```diff
@@ -69,6 +71,7 @@ public SessionRepresentation(
             @JsonProperty("userAgent") Optional<String> userAgent,
             @JsonProperty("clientInfo") Optional<String> clientInfo,
             @JsonProperty("clientTags") Set<String> clientTags,
+            @JsonProperty("resourceEstimate") Optional<ResourceEstimates> resourceEstimate,
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/SessionRepresentation.java]: Is this only `Optional` for backwards compatibility?  If so, how about we change the assignment line below to:
```
this.resourceEstimate = requireNonNull(resourceEstimate, "resourceEstimate is null").orElse(new ResourceEstimate());
```
In generally, I find it weird to have an Optional of a thing where every nested field is Optional.

**Pull request closed without merging**